File: backend/internal/chess/fen.go
package chess

import (
	"fmt"
	"strconv"
	"strings"
)

const StartFEN = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"

var fenToType = map[byte]Piece{'k': King, 'q': Advisor, 'b': Elephant, 'n': Horse, 'r': Rook, 'p': Pawn}
var typeToFEN = map[Piece]byte{King: 'k', Advisor: 'q', Elephant: 'b', Horse: 'n', Rook: 'r', Pawn: 'p'}

func ParseFEN(fen string) (*Position, error) {
	parts := strings.Fields(fen)
	if len(parts) < 1 {
		return nil, fmt.Errorf("empty FEN")
	}
	rows := strings.Split(parts[0], "/")
	if len(rows) != 8 {
		return nil, fmt.Errorf("FEN must have 8 ranks")
	}
	p := &Position{Turn: Red, EnPassant: -1, Fullmove: 1}
	kings := map[Color]int{}
	for i, row := range rows {
		f := 0
		for j := 0; j < len(row); j++ {
			ch := row[j]
			if ch >= '1' && ch <= '8' {
				f += int(ch - '0')
				continue
			}
			t, ok := fenToType[ch|0x20]
			if !ok || f >= 8 {
				return nil, fmt.Errorf("invalid FEN rank %q", row)
			}
			piece := t
			if ch == ch|0x20 {
				piece = -t
			}
			p.Board[Sq(f, 7-i)] = piece
			if t == King {
				kings[piece.Color()]++
			}
			f++
		}
		if f != 8 {
			return nil, fmt.Errorf("invalid FEN rank %q", row)
		}
	}
	if kings[Red] != 1 || kings[Black] != 1 {
		return nil, fmt.Errorf("FEN must have exactly one king per side")
	}
	if len(parts) > 1 && parts[1] == "b" {
		p.Turn = Black
	}
	if len(parts) > 2 && parts[2] != "-" {
		p.Castling = parts[2]
	}
	if len(parts) > 3 && parts[3] != "-" {
		p.EnPassant = Sq(int(parts[3][0]-'a'), int(parts[3][1]-'1'))
	}
	if len(parts) > 4 {
		p.Halfmove, _ = strconv.Atoi(parts[4])
	}
	if len(parts) > 5 {
		p.Fullmove, _ = strconv.Atoi(parts[5])
	}
	return p, nil
}
func (p *Position) FEN() string {
	var b strings.Builder
	for r := 7; r >= 0; r-- {
		empty := 0
		for f := 0; f < 8; f++ {
			x := p.Board[Sq(f, r)]
			if x == Empty {
				empty++
				continue
			}
			if empty > 0 {
				fmt.Fprint(&b, empty)
				empty = 0
			}
			ch := typeToFEN[x.Type()]
			if x > 0 {
				ch -= 32
			}
			b.WriteByte(ch)
		}
		if empty > 0 {
			fmt.Fprint(&b, empty)
		}
		if r > 0 {
			b.WriteByte('/')
		}
	}
	turn := "w"
	if p.Turn == Black {
		turn = "b"
	}
	rights := p.Castling
	if rights == "" {
		rights = "-"
	}
	ep := "-"
	if p.EnPassant >= 0 {
		ep = fmt.Sprintf("%c%d", 'a'+FileOf(p.EnPassant), RankOf(p.EnPassant)+1)
	}
	return fmt.Sprintf("%s %s %s %s %d %d", b.String(), turn, rights, ep, p.Halfmove, p.Fullmove)
}

File: frontend/src/chess/fen.ts
import type { Piece, PieceType, Position } from './types'
import { FILES, RANKS, sq } from './types'

export const START_FEN = 'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1'
const FEN_TO_TYPE: Record<string, PieceType> = { k: 'k', q: 'a', b: 'e', n: 'h', r: 'r', p: 'p', c: 'c' }
const TYPE_TO_FEN: Record<PieceType, string> = { k: 'k', a: 'q', e: 'b', h: 'n', r: 'r', p: 'p', c: 'q' }

export function fromFEN(fen: string): Position {
  const parts = fen.trim().split(/\s+/)
  const rows = parts[0]?.split('/') ?? []
  if (rows.length !== RANKS) throw new Error(`invalid FEN: ${fen}`)
  const board: (Piece | null)[] = new Array(FILES * RANKS).fill(null)
  rows.forEach((row, i) => {
    let file = 0
    for (const ch of row) {
      if (/\d/.test(ch)) file += Number(ch)
      else {
        const type = FEN_TO_TYPE[ch.toLowerCase()]
        if (!type || file >= 8) throw new Error(`invalid FEN: ${fen}`)
        board[sq(file++, 7 - i)] = { color: ch === ch.toUpperCase() ? 'r' : 'b', type }
      }
    }
    if (file !== 8) throw new Error(`invalid FEN rank: ${row}`)
  })
  const ep = parts[3] && parts[3] !== '-' ? sq(parts[3].charCodeAt(0) - 97, Number(parts[3][1]) - 1) : null
  return { board, turn: parts[1] === 'b' ? 'b' : 'r', castling: parts[2] === '-' ? '' : (parts[2] ?? ''), enPassant: ep, halfmove: Number(parts[4] ?? 0), fullmove: Number(parts[5] ?? 1) }
}

export function toFEN(pos: Position): string {
  const rows: string[] = []
  for (let rank = 7; rank >= 0; rank--) {
    let row = '', empty = 0
    for (let file = 0; file < 8; file++) {
      const p = pos.board[sq(file, rank)]
      if (!p) { empty++; continue }
      if (empty) { row += empty; empty = 0 }
      const ch = TYPE_TO_FEN[p.type]
      row += p.color === 'r' ? ch.toUpperCase() : ch
    }
    rows.push(row + (empty || ''))
  }
  const ep = pos.enPassant == null ? '-' : String.fromCharCode(97 + pos.enPassant % 8) + (Math.floor(pos.enPassant / 8) + 1)
  return `${rows.join('/')} ${pos.turn === 'r' ? 'w' : 'b'} ${pos.castling || '-'} ${ep} ${pos.halfmove ?? 0} ${pos.fullmove ?? 1}`
}
